### ISSUE TITLE
Key Within Populated Data

### ISSUE THREAD

**Issue by DeividasK** [prescottprue/react-redux-firebase#40]
Is there a way to leave the ID in the populated object as opposed to it being replaced by the data?

This is the data:
```js
goals: {
  text: 'Something'
  user: 'userId'
}
users: {
  userId: { online: true }
}
```

This is the outcome of populate:
```js
goals: {
  text: 'Something',
  user: { online: true }
}
```

Is there a way to make the outcome as follows:
```js
goals: {
  text: 'Something',
  user: { user: 'userId', online: true }
}
```
or
```js
goals: {
  text: 'Something',
  user: 'userId',
  userData: { online: true }
}
```

The first option would be to include the key used in the populated object. The second option would be to specify the target child for populate values.

**Comment by prescottprue** [comment]
There is not currently a way to do this, but it is a great suggestion.

I like the idea of setting it to another parameter, since if you set a parameter within the object (like user parameter within user), you could overwrite data that is within the populated object.

For now though you can use `populatedDataToJS` (populated data) and `dataToJS`(un populated data) both in your connect function and pass them to different props. This shouldn't cause any extra calls to Firebase since the data is already within redux.

Just to confirm, have you updated to [`v1.2.0`](https://github.com/prescottprue/react-redux-firebase/releases/tag/v1.2.0)?

**Comment by DeividasK** [comment]
Yes, I have updated :)

**Comment by prescottprue** [comment]
## For Now
Since you are on `v1.2.0`, populated data is normalized into redux, so you can do your own version of `populatedDataToJS` like so (code not tested):

```js
import { mapValues } from 'lodash'

const populate = { child: 'user', root: 'users' }

@firebaseConnect(
  ({ params }) => ([
    { path: 'goals', populates: [populate] }
  ])
)
@connect(
  ({ firebase }, { params }) => ({
    goals: mapValues(dataToJS(firebase, 'goals'), (child) => {
      // no matching child parameter (no user property on goal)
      if (!child[populate.child]) {
        return child
      }
      // no matching child does not exist in root list (no user matching user property)
      if (!dataToJS(firebase, `${populate.root}/${child[populate.child]}`)) {
        return child
      }
      // populate child
      return {
        ...child,
        key: child[populate.child], // attach key to new object
        [populate.child]: dataToJS(firebase, `${populate.root}/${child[populate.child]}`)
      }
    })
  })
)
```


I was considering making this be the default for the library (attaching key to any populated object), but then there was a user that brought up the fact that it would override the "key" parameter of an object if it existed. For that reason, this is not the default currently.

## Future Plans
In the future, would you expect this key property to be set by default or would you want to enable it within the populate setting? I personally feel that including it by default could be dangerous (unless there is a setting to change what property the key is set to), but it also provides the easiest usage.

If it did need to be activated via a setting, do you have an idea of what you would want the syntax to look like?

What about like this:
```js
// keyProp is the property of populated object that will contain the key
const populate = { child: 'user', root: 'users', keyProp: 'key' }
@firebaseConnect(
  ({ params }) => ([
    { path: 'goals', populates: [ populate ] }
  ])
)
@connect(
  ({ firebase }, { params }) => ({
    goals: populatedDataToJS(firebase, 'goals', [ populate ]),
  })
)
```

Which would yield something like:

```js
goals: {
  $goalKey : {
    text: 'Something',
    user: { key: 'userId', online: true }
  }
}
```

**Comment by DeividasK** [comment]
Thanks for the custom `populatedDataToJS` version. I'll try that tomorrow.

In the future I think the option that you suggested would be the best way forward. I think in most cases retaining the key is not necessary, so doing this by default may bring more problems than it would solve.  Simply adding `keyProp` like you suggested would be more than enough in cases where one wants to retain the key.

Since having a custom query solves my problem, not sure if this issue should be closed or left for the discussions regarding the future implementation.

Thanks for all the great work!

**Comment by prescottprue** [comment]
The [`v1.3.0-alpha` release](https://github.com/prescottprue/react-redux-firebase/releases/edit/v1.3.0-alpha) includes the start of this functionality.

Let me know if you get a chance to give it a try.

**Comment by prescottprue** [comment]
This has been confirmed and is available in all`v1.3.0-*` releases. Will be included in `v1.3.0` and future versions.
